all: Use Go 1.16 in publish_artifacts and install_golang jobs in Circle

bartekn opened the pull request.

leighmcculloch commented on .circleci/config.yml:
```diff
@@ -385,10 +385,7 @@ jobs:
-      # Debian Stretch is currently not available. See
-      # https://github.com/docker-library/golang/issues/344.
-      - image: golang:1.14.15-stretch
+      - image: golang:1.16
```
As it turns out they just released yesterday stretch variants, so there's that option if you wanted to use it. But I'm 💯 for progressing to Buster based builds if we're able. We should look forward rather than backwards. https://github.com/docker-library/golang/pull/361

leighmcculloch reviewed (approved):
Also need to update changelogs, READMEs, etc that state what the officially supported versions are to add 1.16 and drop 1.14, otherwise 🎉.

paulbellamy commented:
Might be nice to update the go.mod as well so everything is in sync?

bartekn commented:
@leighmcculloch @paulbellamy any idea why there's a [difference](https://app.circleci.com/pipelines/github/stellar/go/7324/workflows/0005e4b0-3e2c-4056-9cb6-612f17e578d6/jobs/36771) in `go.mod` when I update it to contain `go 1.16`?
```diff
diff --git a/go.mod b/go.mod
index 5869a7f62..e8968f811 100644
--- a/go.mod
+++ b/go.mod
@@ -30,7 +30,6 @@ require (
 	github.com/inconshreveable/mousetrap v1.0.0 // indirect
 	github.com/jarcoal/httpmock v0.0.0-20161210151336-4442edb3db31
 	github.com/jmoiron/sqlx v1.2.0
-	github.com/k0kubun/colorstring v0.0.0-20150214042306-9440f1994b88 // indirect
 	github.com/klauspost/compress v0.0.0-20161106143436-e3b7981a12dd // indirect
 	github.com/klauspost/cpuid v0.0.0-20160302075316-09cded8978dc // indirect
 	github.com/klauspost/crc32 v0.0.0-20161016154125-cb6bfca970f6 // indirect
```

leighmcculloch commented:
I don't know why it is being removed, but if you run `go mod why -m <module>` it should show you what import is pulling it in, and it might be that the import was captured in a transient dependencies module file and there was a bug and it wasn't supposed to be here. My only concern would be if our dependencies are changing, if `go.list` is also changing. 

`go.mod` stability is only guaranteed for a single Go release. You can remove `check_code_1_15` too because of that.

leighmcculloch commented:
Oh wait, go.list is changing. 🤔 

https://github.com/stellar/go/pull/3446/files#diff-c5cdb981e916e8188a50107f8f2465b9b4aabf3984b86c48eadde294c5915d68R59-R60

bartekn commented:
> go.mod stability is only guaranteed for a single Go release. You can remove check_code_1_15 too because of that.

That was it, I realized that I got 1.15 or 1.16 failure depending what I've changed. I removed `check_code_1_15` job.

leighmcculloch reviewed (approved):
One typo below. We usually also update the CHANGELOGs, but it is tedious and I think we could probably stop doing that since we focus on releasing binaries.

Regarding the dependency disappearing, this is the `go mod why` when running on Go 1.15:
```
# github.com/k0kubun/colorstring
github.com/stellar/go/support/http/httptest
gopkg.in/gavv/httpexpect.v1
github.com/yudai/gojsondiff
github.com/yudai/gojsondiff.test
github.com/yudai/pp
github.com/yudai/pp.test
github.com/k0kubun/colorstring
```

Looks like we were only dependent on it through a dependencies tests, and looks like that behavior was reported as a bug that was fixed in Go 1.16: https://github.com/golang/go/issues/40799. 🎉 

LGTM

ire-and-curses commented on services/horizon/internal/docs/admin.md:
```diff
@@ -34,7 +34,7 @@ To test the installation, simply run `horizon --help` from a terminal.  If the h
 
 - A unix-like operating system with the common core commands (cp, tar, mkdir, bash, etc.)
-- A compatible distribution of Go (Go 1.14 or later)
+- A compatible distribution of Go (Go 1.16or later)
```
I don't mind dropping the references in the Changelog, if we mention where you can find this info elsewhere. I agree it's tedious.

bartekn replied:
OK, I dropped references to specific versions.

leighmcculloch replied:
Wait, this should be 1.15 or later. We normally support for he last two versions. Or is the plan to only support the most recent?

bartekn commented:
> Regarding the dependency disappearing, this is the go mod why when running on Go 1.15:

Thanks for explanations, makes sense!

leighmcculloch commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/stellar/go
 
-go 1.13
+go 1.16
```
This should be 1.15 because we support the last two versions.

leighmcculloch commented:
Root readme needs updating

bartekn commented:
OK, to sum it up in https://github.com/stellar/go/pull/3446/commits/d0e81c6d4ebcb56d0d0498a034dd71f2b303d1ae I update `go.mod` Go version to 1.15 and removed all references to specific Golang version in all `.md` files, instead I added a note about supporting two latest versions.

leighmcculloch reviewed (approved):
🎉
